### PR TITLE
Set Expressions (with Chainable Format) for Arithmetic and Relational Operations

### PR DESCRIPTION
**Pull request by gadavidd**
**Set Expressions (with Chainable Format) for Arithmetic and Relational Operations**

I've been working with `SetVar` in `choco-Solver`. Unlike other types of variables, such as `RealVar` or `IntVar`, there was no direct way to compose operations such as equals (eq), not equals (ne), contains, not contains (nc), subset (subSet), union, or intersection directly in the model definition. While propagators allow for constructing such constraints, it was not as intuitive as it is for other variable types like `IntVar` and `RealVar`. These features are extremely useful in facilitating the composition of constraints.

Thus, I've worked on adding this feature for `SetVar`, based on the operations already available for other variable types. For example:

```java
realVar.eq(3).post();
```

However, such expressions were not possible for `SetVar`:

```java
setVar.eq(1, 2, 3).post();  // Not available... until now!
```

**Improvement**

I've implemented a new expression framework for `SetVar`, allowing the use of familiar operations directly on set variables. This change introduces support for the following operations:

Relational Operations:

- `eq` (Equals): `setA.eq(1, 2, 3)` or `setA.eq(setB)`.
- `ne` (Not Equals): `setA.ne(1, 2)` or `setA.ne(setB)`.
- `contains`: `setA.contains(1, 2)`.
- `nc` (notContains): `setA.notContains(1, 2)`.
- `subSet`: `setA.subSet(1, 2, 3)` or `setA.subSet(setB)`.

Arithmetic Set Operations:

`union`: `setA.union(setB)`
`intersection`: `setA.intersection(setB)`

This enhancement allows for the expression of complex set relations with simplified syntax, enabling combinations such as:

```java
setA.eq(setB.union(setC.intersection(setD))).post();
```

**Example Usage:**
```java
Model model = new Model();

// Define set variables
SetVarImplArExpression setA = new SetVarImplArExpression(model, "setA", new int[]{}, new int[]{0, 1, 2, 3});
SetVarImplArExpression setB = new SetVarImplArExpression(model, "setB", new int[]{}, new int[]{1, 2, 3});
SetVarImplArExpression setC = new SetVarImplArExpression(model, "setC", new int[]{}, new int[]{2, 3, 4, 5});

// Compose set expressions
setA.eq(setB.union(setC)).post();
model.getSolver().propagate();

// Assert the expected values in setA
assertTrue(Arrays.equals(setA.getUB().toArray(), new int[]{1, 2, 3, 4, 5}));
```

**Internal Changes:**

This change is implemented through three main components:

- Unary Set Expressions (`UnReSetExpression`): Handle relations between a set and a list of values, like `setVar.eq(1,2,3).post()`.
- Binary Set Expressions (`BiReSetExpression`): Manage operations between two `SetVar` instances, such as `setA.eq(setB).post()`.
- Binary Arithmetic Expressions (`BiArSetExpression`): Handle arithmetic operations like unions and intersections, e.g., `setA.eq(setB.union(setC)).post()`.

A new interface, `SetExpression`, inspired by other Choco-Solver expressions, allows `SetVarExpression` to be used in conjunction with these expression composition objects (`UnReSetExpression`, `BiReSetExpression`, and `BiArSetExpression`).

**Backward Compatibility:**

This improvement is fully backward-compatible with existing constraints and models. It adds considerable flexibility when working with `SetVar`, and I believe it will make modeling with sets much more intuitive.

**Future Improvements:**

I believe that additional operations, such as disjoint, could also be implemented to further enhance set operations in Choco-Solver.

Finally, I've included a comprehensive suite of tests in `SetVarExpressionTest`, which I hope will be useful to validate this enhancement.

I would greatly appreciate any feedback or suggestions for optimizations, as well as any edge cases that could further improve this approach!

### CODE REVIEW

**Comment by cprudhom** [comment]
Thank you for the PR. 
It's in my TODOs but I'd like to finish another task before starting on this one, which I think is absolutely relevant.

File: solver/src/main/java/org/chocosolver/solver/expression/discrete/set/BiReSetExpression.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    public Model getModel() {
+        return this.model;
+    }
+
+    @Override
+    public BoolVar boolVar() {
+        if (result == null) {
+            switch (op) {
+                case SUBSET:
+                    result = model.subsetEq(xSet, ySet).reify().eq(TRUE).boolVar();
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/expression/discrete/set/BiReSetExpression.java:56]: The expected expression is : `(xSet \in ySet) \impl result` but you defined: `((xSet \in ySet) \impl True) \imp result`.
You can simplify the expression to:
```java
result = model.subsetEq(xSet, ySet).reify();
```

File: solver/src/main/java/org/chocosolver/solver/expression/discrete/set/BiReSetExpression.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+import org.chocosolver.solver.Model;
+import org.chocosolver.solver.constraints.Constraint;
+import org.chocosolver.solver.expression.discrete.relational.ReExpression;
+import org.chocosolver.solver.variables.BoolVar;
+import org.chocosolver.solver.variables.IntVar;
+import org.chocosolver.solver.variables.SetVar;
+
+import java.util.HashSet;
+
+import static org.chocosolver.solver.constraints.real.Ibex.TRUE;
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/expression/discrete/set/BiReSetExpression.java:13]: `Ibex.TRUE` is a constant for continuous constraint programming and should not be used in Set expressions

File: solver/src/main/java/org/chocosolver/solver/expression/discrete/set/BiReSetExpression.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+    @Override
+    public BoolVar boolVar() {
+        if (result == null) {
+            switch (op) {
+                case SUBSET:
+                    result = model.subsetEq(xSet, ySet).reify().eq(TRUE).boolVar();
+                    break;
+                case EQ:
+                    result = model.allEqual(xSet, ySet).reify().eq(TRUE).boolVar();
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/expression/discrete/set/BiReSetExpression.java:59]: same as l.55

File: solver/src/main/java/org/chocosolver/solver/expression/discrete/set/BiReSetExpression.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+                    break;
+                case NE:
+                    result = model.allDifferent(xSet, ySet).reify().eq(TRUE).boolVar();
+                    break;
+                case CONTAINS:
+                    result = model.subsetEq(ySet, xSet).reify().eq(TRUE).boolVar();
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/expression/discrete/set/BiReSetExpression.java:65]: same as l.55

File: solver/src/main/java/org/chocosolver/solver/expression/discrete/set/BiReSetExpression.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        if (result == null) {
+            switch (op) {
+                case SUBSET:
+                    result = model.subsetEq(xSet, ySet).reify().eq(TRUE).boolVar();
+                    break;
+                case EQ:
+                    result = model.allEqual(xSet, ySet).reify().eq(TRUE).boolVar();
+                    break;
+                case NE:
+                    result = model.allDifferent(xSet, ySet).reify().eq(TRUE).boolVar();
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/expression/discrete/set/BiReSetExpression.java:62]: same as l.55

File: solver/src/main/java/org/chocosolver/solver/expression/discrete/set/BiReSetExpression.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+                    break;
+                case CONTAINS:
+                    result = model.subsetEq(ySet, xSet).reify().eq(TRUE).boolVar();
+                    break;
+                case NC:
+                    result = model.disjoint(ySet, xSet).reify().eq(TRUE).boolVar();
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/expression/discrete/set/BiReSetExpression.java:68]: same as l.55

File: solver/src/main/java/org/chocosolver/solver/expression/discrete/set/SetExpression.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
+package org.chocosolver.solver.expression.discrete.set;
+
+import org.chocosolver.solver.Model;
+import org.chocosolver.solver.constraints.Constraint;
+import org.chocosolver.solver.expression.discrete.relational.ReExpression;
+import org.chocosolver.solver.expression.discrete.relational.UnCReExpression;
+import org.chocosolver.solver.variables.SetVar;
+
+public interface SetExpression {
+    default SetExpression union(SetExpression y){
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/expression/discrete/set/SetExpression.java:10]: I suppose we can have JavaDoc here

File: solver/src/main/java/org/chocosolver/solver/expression/discrete/set/BiReSetExpression.java [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+                default:
+                    throw new UnsupportedOperationException("Unsupported SetOperator: " + op);
+            }
+        }
+        return result;
+    }
+
+    @Override
+    public void extractVar(HashSet<IntVar> variables) {
+    }
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/expression/discrete/set/BiReSetExpression.java:79]: I would throw an exception here, to be safe

File: solver/src/main/java/org/chocosolver/solver/expression/discrete/set/UnReSetExpression.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    private SetExpression x;
+    private int[] y;
+    private Model model;
+    private SetVar xSet, ySet;
+
+    private Constraint constraint;
+
+    private BoolVar result;
+
+    public UnReSetExpression(SetOperator operator, SetExpression x, int... y) {
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/expression/discrete/set/UnReSetExpression.java:37]: I'm not sure to understand why this class is needed, everything can be done with `BiReSetExpression`, no?

File: solver/src/main/java/org/chocosolver/solver/variables/impl/SetVarExpression.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+ * This class extends SetVar to allow arithmetic and relational operations to be performed
+ * directly on set variables, facilitating composition of constraints and complex expressions
+ * such as union, intersection, equality, and subset relations.
+ *
+ * Example:
+ * setA.eq(1,2,3).post();
+ * setA.union(setB).post();
+ */
+
+public class SetVarExpression extends SetVarImpl implements SetExpression {
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/variables/impl/SetVarExpression.java:20]: I don't think this class is useful, can't SetVar implement SetExpression directly?

**Review by cprudhom — CHANGES_REQUESTED** [review]

File: solver/src/main/java/org/chocosolver/solver/variables/impl/SetVarExpression.java (hunk 10 above)
Comment (gadavidd) [reply solver/src/main/java/org/chocosolver/solver/variables/impl/SetVarExpression.java:20]: I initially considered modifying `SetVar` directly, but this approach introduced conflicts due to method inheritance ( `getModel()` ) and could lead to potential breaking changes. Instead, `SetVarExpression` provides a clean abstraction without modifying core components. However, we can consider this change, though it may impact all classes that inherit from SetVar. Would that be a concern?

I initially thought that `SetVarExpression` would only replace `SetVarImpl`, reducing the overall impact. However, I'm open to whatever is best for choco-solver

File: solver/src/main/java/org/chocosolver/solver/expression/discrete/set/BiReSetExpression.java (hunk 1 above)
Comment (gadavidd) [reply solver/src/main/java/org/chocosolver/solver/expression/discrete/set/BiReSetExpression.java]: Updated as suggested

File: solver/src/main/java/org/chocosolver/solver/expression/discrete/set/BiReSetExpression.java (hunk 2 above)
Comment (gadavidd) [reply solver/src/main/java/org/chocosolver/solver/expression/discrete/set/BiReSetExpression.java]: Updated as suggested

File: solver/src/main/java/org/chocosolver/solver/expression/discrete/set/SetExpression.java (hunk 7 above)
Comment (gadavidd) [reply solver/src/main/java/org/chocosolver/solver/expression/discrete/set/SetExpression.java]: Updated as suggested. See if it is in accordance with the added documentation, if anything I can change it again. I tried to follow the standard adopted in the project.

File: solver/src/main/java/org/chocosolver/solver/expression/discrete/set/BiReSetExpression.java (hunk 8 above)
Comment (gadavidd) [reply solver/src/main/java/org/chocosolver/solver/expression/discrete/set/BiReSetExpression.java:77]: Exception added

File: solver/src/main/java/org/chocosolver/solver/expression/discrete/set/UnReSetExpression.java (hunk 9 above)
Comment (gadavidd) [reply solver/src/main/java/org/chocosolver/solver/expression/discrete/set/UnReSetExpression.java:37]: I tried to separate the classes according to responsibility. One for handling the relationship between set and literal value and another for set with set. But I can join them.

File: solver/src/main/java/org/chocosolver/solver/variables/impl/SetVarExpression.java (hunk 10 above)
Comment (gadavidd) [reply solver/src/main/java/org/chocosolver/solver/variables/impl/SetVarExpression.java:20]: Also, I'm sorry, I think it would have been appropriate to have sent you the changes already considering version 5.0.0-beta.1, I will do this in the next corrections.

**Comment by cprudhom** [comment]
Now in #1171
